### PR TITLE
Chiral Atom and Bond Restraints

### PR DESCRIPTION
**Pull request by proteneer**
Implements the GPU backend of chiral atom restraints.

### CODE REVIEW

**Comment by proteneer** [comment]
Note to self: add a fast return force constant is zero.

**Comment by proteneer** [comment]
Note to self 2: bitwise symmetry?

File: tests/test_bonded.py [hunk 1]
```diff
@@ -352,3 +352,57 @@ def test_empty_potentials(self):
...
+            restr_idxs.append(np.random.choice(atom_idxs, size=4, replace=False))
+        restr_idxs = np.array(restr_idxs, dtype=np.int32) if n_restraints else np.zeros((0, 4), dtype=np.int32)
+
+        lamb = 0.0
+        box = np.eye(3) * 100
+
+        relative_tolerance_at_precision = {np.float64: 1e-9, np.float32: 1e-6}
+
+        for precision, rtol in relative_tolerance_at_precision.items():
+            test_potential = potentials.ChiralAtomRestraint(restr_idxs)
```
Comment (badisa) [inline tests/test_bonded.py:376]: Would be good to add a test to verify the shape/values of the restraint idxs

File: timemachine/cpp/src/chiral_bond_restraint.cu [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+template <typename RealType>
+ChiralBondRestraint<RealType>::ChiralBondRestraint(const std::vector<int> &idxs, const std::vector<int> &signs)
+    : R_(idxs.size() / 4) {
+
+    if (idxs.size() % 4 != 0) {
+        throw std::runtime_error("idxs.size() must be exactly 4*R!");
+    }
+
+    if (idxs.size() / 4 != signs.size()) {
```
Comment (badisa) [inline timemachine/cpp/src/chiral_bond_restraint.cu:18]: nit: 
```suggestion
    if (R_ != signs.size()) {
```

File: timemachine/cpp/src/kernels/k_chiral_restraint.cuh [hunk 3]
```diff
@@ -0,0 +1,177 @@
+#include "../fixed_point.hpp"
+#include "chiral_utils.cuh"
+#include "k_fixed_point.cuh"
+
+template <typename RealType>
+void __global__ k_chiral_atom_restraint(
+    const int R, // number of restraints
+    const double *__restrict__ coords,
+    const double *__restrict__ params, // [r]
+    const int *__restrict__ idxs,      // [r, 2]
```
Comment (badisa) [inline timemachine/cpp/src/kernels/k_chiral_restraint.cuh:10]: Is this [r, 4]?

Comment (badisa) [reply timemachine/cpp/src/kernels/k_chiral_restraint.cuh:10]: Might be clearer to be [R] to reference the argument.

File: timemachine/cpp/src/chiral_atom_restraint.hpp [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+template <typename RealType> class ChiralAtomRestraint : public Potential {
+
+private:
+    int *d_idxs_;
+
+    const int R_;
+
+public:
+    ChiralAtomRestraint(const std::vector<int> &idxs // [R, 4]
+    );
```
Comment (mcwitt) [inline timemachine/cpp/src/chiral_atom_restraint.hpp:17]: nit: weird formatting

File: timemachine/cpp/src/chiral_atom_restraint.cu [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+template <typename RealType>
+ChiralAtomRestraint<RealType>::ChiralAtomRestraint(const std::vector<int> &idxs) : R_(idxs.size() / 4) {
+
+    if (idxs.size() % 4 != 0) {
+        throw std::runtime_error("idxs.size() must be exactly 4*k!");
+    }
+
+    gpuErrchk(cudaMalloc(&d_idxs_, R_ * 4 * sizeof(*d_idxs_)));
+    gpuErrchk(cudaMemcpy(d_idxs_, &idxs[0], R_ * 4 * sizeof(*d_idxs_), cudaMemcpyHostToDevice));
```
Comment (mcwitt) [inline timemachine/cpp/src/chiral_atom_restraint.cu:18]: nit: should we switch to `DeviceBuffer` for new code?

File: timemachine/cpp/src/kernels/chiral_utils.cuh [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+    __device__ Vector<RealType> operator-(Vector<RealType> const &obj) const {
+        return {x - obj.x, y - obj.y, z - obj.z};
+    }
+
+    // unary minus
+    __device__ Vector<RealType> operator-() const { return {-x, -y, -z}; }
+};
+
+template <typename RealType> __device__ Vector<RealType> mul_vm(Vector<RealType> v, Matrix<RealType> m) {
```
Comment (mcwitt) [inline timemachine/cpp/src/kernels/chiral_utils.cuh:44]: nit: would it be more conventional to have the matrix as the left argument? (another idea: could be a method of `Matrix`, so you could do `mat.dot(vec)`)

File: timemachine/cpp/src/kernels/chiral_utils.cuh [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+template <typename RealType> __device__ Vector<RealType> cross_product(Vector<RealType> a, Vector<RealType> b) {
+
+    RealType cx = a.y * b.z - a.z * b.y;
+    RealType cy = a.z * b.x - a.x * b.z;
+    RealType cz = a.x * b.y - a.y * b.x;
+
+    return Vector<RealType>({cx, cy, cz});
+}
+
+template <typename RealType> __device__ Vector<RealType> difference(Vector<RealType> a, Vector<RealType> b) {
```
Comment (mcwitt) [inline timemachine/cpp/src/kernels/chiral_utils.cuh:62]: Is this redundant with `operator-` defined above? (also seems unused)

File: timemachine/cpp/src/kernels/chiral_utils.cuh [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    return Vector({a.x - b.x, a.y - b.y, a.z - b.z});
+}
+
+template <typename RealType>
+__device__ void cross_jac(Vector<RealType> a, Vector<RealType> b, Matrix<RealType> &grad_a, Matrix<RealType> &grad_b) {
```
Comment (mcwitt) [inline timemachine/cpp/src/kernels/chiral_utils.cuh:67]: nit: would `jac_a` be a more appropriate name here (vs. `grad_a`)?

**Review by mcwitt — APPROVED** [review]
Looks good, mostly nit comments!

**Review by badisa — APPROVED** [review]

File: timemachine/cpp/src/wrap_kernels.cpp [hunk 9]
```diff
@@ -749,6 +751,37 @@ template <typename RealType> void declare_harmonic_bond(py::module &m, const cha
...
+        m, pyclass_name.c_str(), py::buffer_protocol(), py::dynamic_attr())
+        .def(
+            py::init([](const py::array_t<int, py::array::c_style> &idxs) {
+                std::vector<int> vec_idxs(idxs.data(), idxs.data() + idxs.size());
+                return new timemachine::ChiralAtomRestraint<RealType>(vec_idxs);
+            }),
+            py::arg("idxs"));
+}
+
+template <typename RealType> void declare_chiral_bond_restraint(py::module &m, const char *typestr) {
```
[the remark this answers is not in the document]
Comment (proteneer) [reply timemachine/cpp/src/wrap_kernels.cpp:768]: Are docstrings in C++ now visible from python? I thought they were still hidden when compiled into a .so file.

Usually the docstrings for these are in the reference jax implementations - not sure where's the best place to put these on the C++ side.

File: timemachine/cpp/src/chiral_atom_restraint.hpp (hunk 4 above)
Comment (proteneer) [reply timemachine/cpp/src/chiral_atom_restraint.hpp:17]: Hm - I think it looked a little different before the pre-commit hook automatically formatted this. Maybe the comment throws it off a little?

File: timemachine/cpp/src/chiral_bond_restraint.cu (hunk 2 above)
Comment (proteneer) [reply timemachine/cpp/src/chiral_bond_restraint.cu:18]: willfix inline for all of these.

File: timemachine/cpp/src/kernels/chiral_utils.cuh (hunk 8 above)
Comment (proteneer) [reply timemachine/cpp/src/kernels/chiral_utils.cuh:67]: sure will fix

File: timemachine/cpp/src/kernels/chiral_utils.cuh [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    grad_a.x0 = 0;
```
Comment (proteneer) [inline timemachine/cpp/src/kernels/chiral_utils.cuh:68]: Ugh I think the C++ formatter turned this into a multiline section. I was pretty sure I'd initially written this in matrix form:

```
grad_a.x0 = 0; grad_a.y0 = b.z; grad_a.z0 = -b.y
...
```

which is much easier to read, is there a way to not break it up?

Comment (proteneer) [reply timemachine/cpp/src/kernels/chiral_utils.cuh:68]: @mcwitt ^

File: timemachine/cpp/src/chiral_atom_restraint.hpp (hunk 4 above)
Comment (mcwitt) [reply timemachine/cpp/src/chiral_atom_restraint.hpp:17]: Ah, sorry, I think it's fine actually. I didn't appreciate that the comment is supposed to be paired with the argument versus the whole declaration.

File: timemachine/cpp/src/kernels/chiral_utils.cuh (hunk 10 above)
Comment (mcwitt) [reply timemachine/cpp/src/kernels/chiral_utils.cuh:68]: Does this work? https://clang.llvm.org/docs/ClangFormatStyleOptions.html#disabling-formatting-on-a-piece-of-code

File: tests/test_bonded.py [hunk 11]
```diff
@@ -352,3 +352,57 @@ def test_empty_potentials(self):
         self.test_periodic_torsion(n_torsions=0)
         self.test_harmonic_angle(n_angles=0)
         self.test_harmonic_bond(n_bonds=0)
+
+    def test_chiral_atom_restraint(self, n_particles=64, n_restraints=35, dim=3):
```
[the remark this answers is not in the document]
Comment (proteneer) [reply tests/test_bonded.py:356]: it's not - but its copy and pasted from another test @maxentile I think suggested to move all constants to a parameter.

File: tests/test_bonded.py [hunk 12]
```diff
@@ -352,3 +352,57 @@ def test_empty_potentials(self):
...
+
+        x = self.get_random_coords(n_particles, dim)
+
+        atom_idxs = np.arange(n_particles)
+        params = np.random.rand(n_restraints).astype(np.float64)
+
+        restr_idxs = []
+        for _ in range(n_restraints):
+            restr_idxs.append(np.random.choice(atom_idxs, size=4, replace=False))
+        restr_idxs = np.array(restr_idxs, dtype=np.int32) if n_restraints else np.zeros((0, 4), dtype=np.int32)
```
[the remark this answers is not in the document]
Comment (proteneer) [reply tests/test_bonded.py:368]: I think its when someone runs this with `n_restraints=0`

File: tests/test_bonded.py [hunk 13]
```diff
@@ -352,3 +352,57 @@ def test_empty_potentials(self):
...
+        lamb = 0.0
+        box = np.eye(3) * 100
+
+        relative_tolerance_at_precision = {np.float64: 1e-9, np.float32: 1e-6}
+
+        for precision, rtol in relative_tolerance_at_precision.items():
+            test_potential = potentials.ChiralBondRestraint(restr_idxs, signs)
+            ref_potential = functools.partial(chiral_restraints.chiral_bond_restraint, idxs=restr_idxs, signs=signs)
+
+            self.compare_forces(x, params, box, [lamb], ref_potential, test_potential, rtol, precision=precision)
```
[the remark this answers is not in the document]
Comment (proteneer) [reply tests/test_bonded.py:408]: I can try to do some validation here - to see if assertions get thrown

File: timemachine/cpp/src/chiral_atom_restraint.cu (hunk 5 above)
Comment (proteneer) [reply timemachine/cpp/src/chiral_atom_restraint.cu:18]: Probably - this was copy and pasted from the `HarmonicBond` code, should we just try to do them all during deboggling week for consistency?

File: timemachine/cpp/src/kernels/chiral_utils.cuh (hunk 6 above)
Comment (proteneer) [reply timemachine/cpp/src/kernels/chiral_utils.cuh:44]: this is a left multiply of a vector * matrix - (note the signature of `_vm`) so probably best to be consistent in its args?

File: timemachine/cpp/src/kernels/chiral_utils.cuh (hunk 7 above)
Comment (proteneer) [reply timemachine/cpp/src/kernels/chiral_utils.cuh:62]: good call, will fix

File: timemachine/cpp/src/kernels/chiral_utils.cuh (hunk 10 above)
Comment (proteneer) [reply timemachine/cpp/src/kernels/chiral_utils.cuh:68]: will test

File: timemachine/cpp/src/wrap_kernels.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (proteneer) [reply timemachine/cpp/src/wrap_kernels.cpp:768]: yeah I think that's simpler

File: tests/test_bonded.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (mcwitt) [reply tests/test_bonded.py:368]: Ah ok -- I thought I had tested this with `n_restraints=0` and got a different error, but it seems to work now so I must've been mistaken.

File: tests/test_bonded.py [hunk 14]
```diff
@@ -352,3 +352,57 @@ def test_empty_potentials(self):
...
+        np.random.seed(125)  # TODO: where should this seed be set?
+
+        x = self.get_random_coords(n_particles, dim)
+
+        atom_idxs = np.arange(n_particles)
+        params = np.random.rand(n_restraints).astype(np.float64)
+
+        restr_idxs = []
+        for _ in range(n_restraints):
+            restr_idxs.append(np.random.choice(atom_idxs, size=4, replace=False))
```
[the remark this answers is not in the document]
Comment (proteneer) [reply tests/test_bonded.py:367]: willfix inline

File: timemachine/cpp/src/kernels/chiral_utils.cuh (hunk 6 above)
Comment (mcwitt) [reply timemachine/cpp/src/kernels/chiral_utils.cuh:44]: Got it, makes sense. I misread the implementation as multiplication of a matrix by a column vector on the right.

**Pull request merged**